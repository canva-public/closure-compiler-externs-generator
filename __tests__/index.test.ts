import {
  createApplyDefaults,
  processLibraries,
  FS,
  Library,
} from '../src/index';
import { Volume, createFsFromVolume } from 'memfs';
import { libraries } from './fixtures/libraries';
import * as path from 'path';

const fixturesDir = path.resolve(__dirname, 'fixtures');

function createVolumeAndFs() {
  const volume = new Volume();
  const fs = createFsFromVolume(volume) as FS;

  return {
    fs,
    volume,
  };
}

function snapshotLibraries(
  libraries: Partial<Library> & { moduleName: string }[],
) {
  const tailoredApplyDefaults = createApplyDefaults(fixturesDir);
  const { volume, fs } = createVolumeAndFs();
  processLibraries(
    '/',
    libraries.map(tailoredApplyDefaults),
    false,
    fs,
    fixturesDir,
  );
  expect(volume.toJSON()).toMatchSnapshot();
}

describe('externs-generator', () => {
  describe('generation', () => {
    it.each([false, true])(
      'produces externs for a given set of modules with debug = %s',
      (debug) => {
        expect.hasAssertions();
        const { volume, fs } = createVolumeAndFs();
        processLibraries('/', libraries, debug, fs, fixturesDir);
        expect(volume.toJSON()).toMatchSnapshot();
      },
    );

    it('for scoped modules', () => {
      expect.hasAssertions();
      snapshotLibraries([
        { moduleName: '@scoped/exports-sugar-esm' },
        { moduleName: 'cjs-named-exports' },
        { moduleName: 'main-implicit' },
        { moduleName: 'untyped-cjs-and-esm' },
        { moduleName: 'untyped-cjs' },
      ]);
    });

    it('for various single-export modules', () => {
      expect.hasAssertions();
      snapshotLibraries([
        { moduleName: 'cjs-named-exports' },
        { moduleName: 'main-implicit' },
        { moduleName: 'untyped-cjs-and-esm' },
        { moduleName: 'untyped-cjs' },
      ]);
    });
  });
});
