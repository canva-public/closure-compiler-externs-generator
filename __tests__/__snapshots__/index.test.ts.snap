// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`externs-generator generation for scoped modules 1`] = `
Object {
  "/index.js": "module.exports.useFoo = () => {};",
  "/index.mjs": "export const useFoo = () => {};",
  "/node_modules/@scoped/exports-multiple-cjs-and-esm/extra/package.json": "{
  \\"name\\": \\"extra\\",
  \\"main\\": \\"./src/extra.js\\",
  \\"types\\": \\"./src/extra.d.ts\\"
}",
  "/node_modules/@scoped/exports-multiple-cjs-and-esm/extra/src/extra.d.ts": "export = Scoped_ExportsMultipleCjs_Extra;
declare namespace Scoped_ExportsMultipleCjs_Extra { function extra(): void; }",
  "/node_modules/@scoped/exports-multiple-cjs-and-esm/extra/src/extra.js": "module.exports.extra = () => {};",
  "/node_modules/@scoped/exports-multiple-cjs-and-esm/extra/src/extra.mjs": "export const extra = () => {};",
  "/node_modules/@scoped/exports-multiple-cjs-and-esm/package.json": "{
  \\"name\\": \\"@scoped/exports-multiple-cjs-and-esm\\",
  \\"main\\": \\"./src/main.js\\",
  \\"exports\\": {
    \\"./\\": {
      \\"import\\": \\"./src/main.mjs\\",
      \\"require\\": \\"./src/main.js\\"
    },
    \\"./extra\\": {
      \\"import\\": \\"./extra/src/extra.mjs\\",
      \\"require\\": \\"./extra/src/extra.js\\"
    }
  },
  \\"types\\": \\"./src/main.d.ts\\"
}",
  "/node_modules/@scoped/exports-multiple-cjs-and-esm/src/main.d.ts": "export = Scoped_ExportsMultipleCJS;
declare namespace Scoped_ExportsMultipleCJS { function main(): void; }",
  "/node_modules/@scoped/exports-multiple-cjs-and-esm/src/main.js": "module.exports.main = () => {};",
  "/node_modules/@scoped/exports-multiple-cjs-and-esm/src/main.mjs": "export const main = () => {};",
  "/node_modules/@scoped/exports-multiple-cjs/extra/package.json": "{
  \\"name\\": \\"extra\\",
  \\"main\\": \\"./src/extra.js\\",
  \\"types\\": \\"./src/extra.d.ts\\"
}",
  "/node_modules/@scoped/exports-multiple-cjs/extra/src/extra.d.ts": "export = Scoped_ExportsMultipleCjs_Extra;
declare namespace Scoped_ExportsMultipleCjs_Extra { function extra(): void; }",
  "/node_modules/@scoped/exports-multiple-cjs/extra/src/extra.js": "module.exports.extra = () => {};",
  "/node_modules/@scoped/exports-multiple-cjs/package.json": "{
  \\"name\\": \\"@scoped/exports-multiple-cjs\\",
  \\"main\\": \\"./src/main.js\\",
  \\"types\\": \\"./src/main.d.ts\\"
}",
  "/node_modules/@scoped/exports-multiple-cjs/src/main.d.ts": "export = Scoped_ExportsMultipleCJS;
declare namespace Scoped_ExportsMultipleCJS { function main(): void; }",
  "/node_modules/@scoped/exports-multiple-cjs/src/main.js": "module.exports.main = () => {};",
  "/node_modules/@scoped/exports-multiple-esm/extra.d.ts": "export function extra(): void;",
  "/node_modules/@scoped/exports-multiple-esm/extra.js": "export function extra() {}",
  "/node_modules/@scoped/exports-multiple-esm/main.d.ts": "export function main(): void;",
  "/node_modules/@scoped/exports-multiple-esm/main.js": "export function main() {}",
  "/node_modules/@scoped/exports-multiple-esm/package.json": "{
  \\"name\\": \\"@scoped/exports-multiple-esm\\",
  \\"type\\": \\"module\\",
  \\"exports\\": {
    \\"./\\": \\"./main.js\\",
    \\"./extra\\": \\"./extra.js\\"
  }
}",
  "/node_modules/@scoped/exports-sugar-esm/main.d.ts": "export function foobar(): void;
export const barfoo: string;",
  "/node_modules/@scoped/exports-sugar-esm/main.js": "export function foobar() {}
export const barfoo = '';",
  "/node_modules/@scoped/exports-sugar-esm/package.json": "{
  \\"name\\": \\"@scoped/exports-sugar-esm\\",
  \\"type\\": \\"module\\",
  \\"exports\\": \\"./main.js\\"
}",
  "/node_modules/@scoped/exports-sugarfree-esm/main.d.ts": "export function foobar(): void;
export const barfoo: string;",
  "/node_modules/@scoped/exports-sugarfree-esm/main.js": "export function foobar() {}
export const barfoo = '';",
  "/node_modules/@scoped/exports-sugarfree-esm/package.json": "{
  \\"name\\": \\"@scoped/exports-sugarfree-esm\\",
  \\"type\\": \\"module\\",
  \\"exports\\": {
    \\"./\\": \\"./main.js\\"
  }
}",
  "/node_modules/@types/untyped-cjs-and-esm/index.d.ts": "export = UntypedCjs;
declare namespace UntypedCjsAndEsm { function useFoo(): void; }",
  "/node_modules/@types/untyped-cjs-and-esm/package.json": "{
  \\"name\\": \\"@types/untyped-cjs-and-esm\\",
  \\"types\\": \\"./index.d.ts\\"
}",
  "/node_modules/@types/untyped-cjs/index.d.ts": "export = UntypedCjs;
declare namespace UntypedCjs { function useFoo(): void; }",
  "/node_modules/@types/untyped-cjs/package.json": "{
  \\"name\\": \\"@types/untyped-cjs\\",
  \\"types\\": \\"./index.d.ts\\"
}",
  "/node_modules/cjs-named-exports/main-types.d.ts": "export function foo(): void;
export function bar(): string;",
  "/node_modules/cjs-named-exports/main.js": "module.exports.foo = () => {};
module.exports.bar = () => '';",
  "/node_modules/cjs-named-exports/package.json": "{
  \\"name\\": \\"cjs-named-exports\\"
}",
  "/node_modules/main-implicit/index.d.ts": "export = foo;
declare function foo(): void;",
  "/node_modules/main-implicit/index.js": "module.exports = function foo() {}",
  "/node_modules/main-implicit/package.json": "{
  \\"name\\": \\"main-implicit\\"
}",
  "/node_modules/typings-synonym/main-types.d.ts": "export function foo(): void;
export function bar(): string;",
  "/node_modules/typings-synonym/main.js": "module.exports.foo = () => {};
module.exports.bar = () => '';",
  "/node_modules/typings-synonym/package.json": "{
  \\"name\\": \\"typings-synonym\\",
  \\"main\\": \\"./main.js\\",
  \\"typings\\": \\"./main-types.d.ts\\"
}",
  "/package.json": "{
  \\"name\\": \\"untyped-cjs-and-esm\\",
  \\"main\\": \\"./main.js\\",
  \\"exports\\": {
    \\"./\\": {
      \\"require\\": \\"./main.js\\",
      \\"import\\": \\"./main.mjs\\"
    }
  }
}",
}
`;

exports[`externs-generator generation for various single-export modules 1`] = `
Object {
  "/index.js": "module.exports.useFoo = () => {};",
  "/index.mjs": "export const useFoo = () => {};",
  "/node_modules/@scoped/exports-multiple-cjs-and-esm/extra/package.json": "{
  \\"name\\": \\"extra\\",
  \\"main\\": \\"./src/extra.js\\",
  \\"types\\": \\"./src/extra.d.ts\\"
}",
  "/node_modules/@scoped/exports-multiple-cjs-and-esm/extra/src/extra.d.ts": "export = Scoped_ExportsMultipleCjs_Extra;
declare namespace Scoped_ExportsMultipleCjs_Extra { function extra(): void; }",
  "/node_modules/@scoped/exports-multiple-cjs-and-esm/extra/src/extra.js": "module.exports.extra = () => {};",
  "/node_modules/@scoped/exports-multiple-cjs-and-esm/extra/src/extra.mjs": "export const extra = () => {};",
  "/node_modules/@scoped/exports-multiple-cjs-and-esm/package.json": "{
  \\"name\\": \\"@scoped/exports-multiple-cjs-and-esm\\",
  \\"main\\": \\"./src/main.js\\",
  \\"exports\\": {
    \\"./\\": {
      \\"import\\": \\"./src/main.mjs\\",
      \\"require\\": \\"./src/main.js\\"
    },
    \\"./extra\\": {
      \\"import\\": \\"./extra/src/extra.mjs\\",
      \\"require\\": \\"./extra/src/extra.js\\"
    }
  },
  \\"types\\": \\"./src/main.d.ts\\"
}",
  "/node_modules/@scoped/exports-multiple-cjs-and-esm/src/main.d.ts": "export = Scoped_ExportsMultipleCJS;
declare namespace Scoped_ExportsMultipleCJS { function main(): void; }",
  "/node_modules/@scoped/exports-multiple-cjs-and-esm/src/main.js": "module.exports.main = () => {};",
  "/node_modules/@scoped/exports-multiple-cjs-and-esm/src/main.mjs": "export const main = () => {};",
  "/node_modules/@scoped/exports-multiple-cjs/extra/package.json": "{
  \\"name\\": \\"extra\\",
  \\"main\\": \\"./src/extra.js\\",
  \\"types\\": \\"./src/extra.d.ts\\"
}",
  "/node_modules/@scoped/exports-multiple-cjs/extra/src/extra.d.ts": "export = Scoped_ExportsMultipleCjs_Extra;
declare namespace Scoped_ExportsMultipleCjs_Extra { function extra(): void; }",
  "/node_modules/@scoped/exports-multiple-cjs/extra/src/extra.js": "module.exports.extra = () => {};",
  "/node_modules/@scoped/exports-multiple-cjs/package.json": "{
  \\"name\\": \\"@scoped/exports-multiple-cjs\\",
  \\"main\\": \\"./src/main.js\\",
  \\"types\\": \\"./src/main.d.ts\\"
}",
  "/node_modules/@scoped/exports-multiple-cjs/src/main.d.ts": "export = Scoped_ExportsMultipleCJS;
declare namespace Scoped_ExportsMultipleCJS { function main(): void; }",
  "/node_modules/@scoped/exports-multiple-cjs/src/main.js": "module.exports.main = () => {};",
  "/node_modules/@scoped/exports-multiple-esm/extra.d.ts": "export function extra(): void;",
  "/node_modules/@scoped/exports-multiple-esm/extra.js": "export function extra() {}",
  "/node_modules/@scoped/exports-multiple-esm/main.d.ts": "export function main(): void;",
  "/node_modules/@scoped/exports-multiple-esm/main.js": "export function main() {}",
  "/node_modules/@scoped/exports-multiple-esm/package.json": "{
  \\"name\\": \\"@scoped/exports-multiple-esm\\",
  \\"type\\": \\"module\\",
  \\"exports\\": {
    \\"./\\": \\"./main.js\\",
    \\"./extra\\": \\"./extra.js\\"
  }
}",
  "/node_modules/@scoped/exports-sugar-esm/main.d.ts": "export function foobar(): void;
export const barfoo: string;",
  "/node_modules/@scoped/exports-sugar-esm/main.js": "export function foobar() {}
export const barfoo = '';",
  "/node_modules/@scoped/exports-sugar-esm/package.json": "{
  \\"name\\": \\"@scoped/exports-sugar-esm\\",
  \\"type\\": \\"module\\",
  \\"exports\\": \\"./main.js\\"
}",
  "/node_modules/@scoped/exports-sugarfree-esm/main.d.ts": "export function foobar(): void;
export const barfoo: string;",
  "/node_modules/@scoped/exports-sugarfree-esm/main.js": "export function foobar() {}
export const barfoo = '';",
  "/node_modules/@scoped/exports-sugarfree-esm/package.json": "{
  \\"name\\": \\"@scoped/exports-sugarfree-esm\\",
  \\"type\\": \\"module\\",
  \\"exports\\": {
    \\"./\\": \\"./main.js\\"
  }
}",
  "/node_modules/@types/untyped-cjs-and-esm/index.d.ts": "export = UntypedCjs;
declare namespace UntypedCjsAndEsm { function useFoo(): void; }",
  "/node_modules/@types/untyped-cjs-and-esm/package.json": "{
  \\"name\\": \\"@types/untyped-cjs-and-esm\\",
  \\"types\\": \\"./index.d.ts\\"
}",
  "/node_modules/@types/untyped-cjs/index.d.ts": "export = UntypedCjs;
declare namespace UntypedCjs { function useFoo(): void; }",
  "/node_modules/@types/untyped-cjs/package.json": "{
  \\"name\\": \\"@types/untyped-cjs\\",
  \\"types\\": \\"./index.d.ts\\"
}",
  "/node_modules/cjs-named-exports/main-types.d.ts": "export function foo(): void;
export function bar(): string;",
  "/node_modules/cjs-named-exports/main.js": "module.exports.foo = () => {};
module.exports.bar = () => '';",
  "/node_modules/cjs-named-exports/package.json": "{
  \\"name\\": \\"cjs-named-exports\\"
}",
  "/node_modules/main-implicit/index.d.ts": "export = foo;
declare function foo(): void;",
  "/node_modules/main-implicit/index.js": "module.exports = function foo() {}",
  "/node_modules/main-implicit/package.json": "{
  \\"name\\": \\"main-implicit\\"
}",
  "/node_modules/typings-synonym/main-types.d.ts": "export function foo(): void;
export function bar(): string;",
  "/node_modules/typings-synonym/main.js": "module.exports.foo = () => {};
module.exports.bar = () => '';",
  "/node_modules/typings-synonym/package.json": "{
  \\"name\\": \\"typings-synonym\\",
  \\"main\\": \\"./main.js\\",
  \\"typings\\": \\"./main-types.d.ts\\"
}",
  "/package.json": "{
  \\"name\\": \\"untyped-cjs-and-esm\\",
  \\"main\\": \\"./main.js\\",
  \\"exports\\": {
    \\"./\\": {
      \\"require\\": \\"./main.js\\",
      \\"import\\": \\"./main.mjs\\"
    }
  }
}",
}
`;

exports[`externs-generator generation produces externs for a given set of modules with debug = false 1`] = `
Object {
  "/index.js": "module.exports.useFoo = () => {};",
  "/index.mjs": "export const useFoo = () => {};",
  "/node_modules/@scoped/exports-multiple-cjs-and-esm/extra/package.json": "{
  \\"name\\": \\"extra\\",
  \\"main\\": \\"./src/extra.js\\",
  \\"types\\": \\"./src/extra.d.ts\\"
}",
  "/node_modules/@scoped/exports-multiple-cjs-and-esm/extra/src/extra.d.ts": "export = Scoped_ExportsMultipleCjs_Extra;
declare namespace Scoped_ExportsMultipleCjs_Extra { function extra(): void; }",
  "/node_modules/@scoped/exports-multiple-cjs-and-esm/extra/src/extra.js": "module.exports.extra = () => {};",
  "/node_modules/@scoped/exports-multiple-cjs-and-esm/extra/src/extra.mjs": "export const extra = () => {};",
  "/node_modules/@scoped/exports-multiple-cjs-and-esm/package.json": "{
  \\"name\\": \\"@scoped/exports-multiple-cjs-and-esm\\",
  \\"main\\": \\"./src/main.js\\",
  \\"exports\\": {
    \\"./\\": {
      \\"import\\": \\"./src/main.mjs\\",
      \\"require\\": \\"./src/main.js\\"
    },
    \\"./extra\\": {
      \\"import\\": \\"./extra/src/extra.mjs\\",
      \\"require\\": \\"./extra/src/extra.js\\"
    }
  },
  \\"types\\": \\"./src/main.d.ts\\"
}",
  "/node_modules/@scoped/exports-multiple-cjs-and-esm/src/main.d.ts": "export = Scoped_ExportsMultipleCJS;
declare namespace Scoped_ExportsMultipleCJS { function main(): void; }",
  "/node_modules/@scoped/exports-multiple-cjs-and-esm/src/main.js": "module.exports.main = () => {};",
  "/node_modules/@scoped/exports-multiple-cjs-and-esm/src/main.mjs": "export const main = () => {};",
  "/node_modules/@scoped/exports-multiple-cjs/extra/package.json": "{
  \\"name\\": \\"extra\\",
  \\"main\\": \\"./src/extra.js\\",
  \\"types\\": \\"./src/extra.d.ts\\"
}",
  "/node_modules/@scoped/exports-multiple-cjs/extra/src/extra.d.ts": "export = Scoped_ExportsMultipleCjs_Extra;
declare namespace Scoped_ExportsMultipleCjs_Extra { function extra(): void; }",
  "/node_modules/@scoped/exports-multiple-cjs/extra/src/extra.js": "module.exports.extra = () => {};",
  "/node_modules/@scoped/exports-multiple-cjs/package.json": "{
  \\"name\\": \\"@scoped/exports-multiple-cjs\\",
  \\"main\\": \\"./src/main.js\\",
  \\"types\\": \\"./src/main.d.ts\\"
}",
  "/node_modules/@scoped/exports-multiple-cjs/src/main.d.ts": "export = Scoped_ExportsMultipleCJS;
declare namespace Scoped_ExportsMultipleCJS { function main(): void; }",
  "/node_modules/@scoped/exports-multiple-cjs/src/main.js": "module.exports.main = () => {};",
  "/node_modules/@scoped/exports-multiple-esm/extra.d.ts": "export function extra(): void;",
  "/node_modules/@scoped/exports-multiple-esm/extra.js": "export function extra() {}",
  "/node_modules/@scoped/exports-multiple-esm/main.d.ts": "export function main(): void;",
  "/node_modules/@scoped/exports-multiple-esm/main.js": "export function main() {}",
  "/node_modules/@scoped/exports-multiple-esm/package.json": "{
  \\"name\\": \\"@scoped/exports-multiple-esm\\",
  \\"type\\": \\"module\\",
  \\"exports\\": {
    \\"./\\": \\"./main.js\\",
    \\"./extra\\": \\"./extra.js\\"
  }
}",
  "/node_modules/@scoped/exports-sugar-esm/main.d.ts": "export function foobar(): void;
export const barfoo: string;",
  "/node_modules/@scoped/exports-sugar-esm/main.js": "export function foobar() {}
export const barfoo = '';",
  "/node_modules/@scoped/exports-sugar-esm/package.json": "{
  \\"name\\": \\"@scoped/exports-sugar-esm\\",
  \\"type\\": \\"module\\",
  \\"exports\\": \\"./main.js\\"
}",
  "/node_modules/@scoped/exports-sugarfree-esm/main.d.ts": "export function foobar(): void;
export const barfoo: string;",
  "/node_modules/@scoped/exports-sugarfree-esm/main.js": "export function foobar() {}
export const barfoo = '';",
  "/node_modules/@scoped/exports-sugarfree-esm/package.json": "{
  \\"name\\": \\"@scoped/exports-sugarfree-esm\\",
  \\"type\\": \\"module\\",
  \\"exports\\": {
    \\"./\\": \\"./main.js\\"
  }
}",
  "/node_modules/@types/untyped-cjs-and-esm/index.d.ts": "export = UntypedCjs;
declare namespace UntypedCjsAndEsm { function useFoo(): void; }",
  "/node_modules/@types/untyped-cjs-and-esm/package.json": "{
  \\"name\\": \\"@types/untyped-cjs-and-esm\\",
  \\"types\\": \\"./index.d.ts\\"
}",
  "/node_modules/@types/untyped-cjs/index.d.ts": "export = UntypedCjs;
declare namespace UntypedCjs { function useFoo(): void; }",
  "/node_modules/@types/untyped-cjs/package.json": "{
  \\"name\\": \\"@types/untyped-cjs\\",
  \\"types\\": \\"./index.d.ts\\"
}",
  "/node_modules/cjs-named-exports/main-types.d.ts": "export function foo(): void;
export function bar(): string;",
  "/node_modules/cjs-named-exports/main.js": "module.exports.foo = () => {};
module.exports.bar = () => '';",
  "/node_modules/cjs-named-exports/package.json": "{
  \\"name\\": \\"cjs-named-exports\\"
}",
  "/node_modules/main-implicit/index.d.ts": "export = foo;
declare function foo(): void;",
  "/node_modules/main-implicit/index.js": "module.exports = function foo() {}",
  "/node_modules/main-implicit/package.json": "{
  \\"name\\": \\"main-implicit\\"
}",
  "/node_modules/typings-synonym/main-types.d.ts": "export function foo(): void;
export function bar(): string;",
  "/node_modules/typings-synonym/main.js": "module.exports.foo = () => {};
module.exports.bar = () => '';",
  "/node_modules/typings-synonym/package.json": "{
  \\"name\\": \\"typings-synonym\\",
  \\"main\\": \\"./main.js\\",
  \\"typings\\": \\"./main-types.d.ts\\"
}",
  "/package.json": "{
  \\"name\\": \\"untyped-cjs-and-esm\\",
  \\"main\\": \\"./main.js\\",
  \\"exports\\": {
    \\"./\\": {
      \\"require\\": \\"./main.js\\",
      \\"import\\": \\"./main.mjs\\"
    }
  }
}",
}
`;

exports[`externs-generator generation produces externs for a given set of modules with debug = true 1`] = `
Object {
  "/index.js": "module.exports.useFoo = () => {};",
  "/index.mjs": "export const useFoo = () => {};",
  "/node_modules/@scoped/exports-multiple-cjs-and-esm/extra/package.json": "{
  \\"name\\": \\"extra\\",
  \\"main\\": \\"./src/extra.js\\",
  \\"types\\": \\"./src/extra.d.ts\\"
}",
  "/node_modules/@scoped/exports-multiple-cjs-and-esm/extra/src/extra.d.ts": "export = Scoped_ExportsMultipleCjs_Extra;
declare namespace Scoped_ExportsMultipleCjs_Extra { function extra(): void; }",
  "/node_modules/@scoped/exports-multiple-cjs-and-esm/extra/src/extra.js": "module.exports.extra = () => {};",
  "/node_modules/@scoped/exports-multiple-cjs-and-esm/extra/src/extra.mjs": "export const extra = () => {};",
  "/node_modules/@scoped/exports-multiple-cjs-and-esm/package.json": "{
  \\"name\\": \\"@scoped/exports-multiple-cjs-and-esm\\",
  \\"main\\": \\"./src/main.js\\",
  \\"exports\\": {
    \\"./\\": {
      \\"import\\": \\"./src/main.mjs\\",
      \\"require\\": \\"./src/main.js\\"
    },
    \\"./extra\\": {
      \\"import\\": \\"./extra/src/extra.mjs\\",
      \\"require\\": \\"./extra/src/extra.js\\"
    }
  },
  \\"types\\": \\"./src/main.d.ts\\"
}",
  "/node_modules/@scoped/exports-multiple-cjs-and-esm/src/main.d.ts": "export = Scoped_ExportsMultipleCJS;
declare namespace Scoped_ExportsMultipleCJS { function main(): void; }",
  "/node_modules/@scoped/exports-multiple-cjs-and-esm/src/main.js": "module.exports.main = () => {};",
  "/node_modules/@scoped/exports-multiple-cjs-and-esm/src/main.mjs": "export const main = () => {};",
  "/node_modules/@scoped/exports-multiple-cjs/extra/package.json": "{
  \\"name\\": \\"extra\\",
  \\"main\\": \\"./src/extra.js\\",
  \\"types\\": \\"./src/extra.d.ts\\"
}",
  "/node_modules/@scoped/exports-multiple-cjs/extra/src/extra.d.ts": "export = Scoped_ExportsMultipleCjs_Extra;
declare namespace Scoped_ExportsMultipleCjs_Extra { function extra(): void; }",
  "/node_modules/@scoped/exports-multiple-cjs/extra/src/extra.js": "module.exports.extra = () => {};",
  "/node_modules/@scoped/exports-multiple-cjs/package.json": "{
  \\"name\\": \\"@scoped/exports-multiple-cjs\\",
  \\"main\\": \\"./src/main.js\\",
  \\"types\\": \\"./src/main.d.ts\\"
}",
  "/node_modules/@scoped/exports-multiple-cjs/src/main.d.ts": "export = Scoped_ExportsMultipleCJS;
declare namespace Scoped_ExportsMultipleCJS { function main(): void; }",
  "/node_modules/@scoped/exports-multiple-cjs/src/main.js": "module.exports.main = () => {};",
  "/node_modules/@scoped/exports-multiple-esm/extra.d.ts": "export function extra(): void;",
  "/node_modules/@scoped/exports-multiple-esm/extra.js": "export function extra() {}",
  "/node_modules/@scoped/exports-multiple-esm/main.d.ts": "export function main(): void;",
  "/node_modules/@scoped/exports-multiple-esm/main.js": "export function main() {}",
  "/node_modules/@scoped/exports-multiple-esm/package.json": "{
  \\"name\\": \\"@scoped/exports-multiple-esm\\",
  \\"type\\": \\"module\\",
  \\"exports\\": {
    \\"./\\": \\"./main.js\\",
    \\"./extra\\": \\"./extra.js\\"
  }
}",
  "/node_modules/@scoped/exports-sugar-esm/main.d.ts": "export function foobar(): void;
export const barfoo: string;",
  "/node_modules/@scoped/exports-sugar-esm/main.js": "export function foobar() {}
export const barfoo = '';",
  "/node_modules/@scoped/exports-sugar-esm/package.json": "{
  \\"name\\": \\"@scoped/exports-sugar-esm\\",
  \\"type\\": \\"module\\",
  \\"exports\\": \\"./main.js\\"
}",
  "/node_modules/@scoped/exports-sugarfree-esm/main.d.ts": "export function foobar(): void;
export const barfoo: string;",
  "/node_modules/@scoped/exports-sugarfree-esm/main.js": "export function foobar() {}
export const barfoo = '';",
  "/node_modules/@scoped/exports-sugarfree-esm/package.json": "{
  \\"name\\": \\"@scoped/exports-sugarfree-esm\\",
  \\"type\\": \\"module\\",
  \\"exports\\": {
    \\"./\\": \\"./main.js\\"
  }
}",
  "/node_modules/@types/untyped-cjs-and-esm/index.d.ts": "export = UntypedCjs;
declare namespace UntypedCjsAndEsm { function useFoo(): void; }",
  "/node_modules/@types/untyped-cjs-and-esm/package.json": "{
  \\"name\\": \\"@types/untyped-cjs-and-esm\\",
  \\"types\\": \\"./index.d.ts\\"
}",
  "/node_modules/@types/untyped-cjs/index.d.ts": "export = UntypedCjs;
declare namespace UntypedCjs { function useFoo(): void; }",
  "/node_modules/@types/untyped-cjs/package.json": "{
  \\"name\\": \\"@types/untyped-cjs\\",
  \\"types\\": \\"./index.d.ts\\"
}",
  "/node_modules/cjs-named-exports/main-types.d.ts": "export function foo(): void;
export function bar(): string;",
  "/node_modules/cjs-named-exports/main.js": "module.exports.foo = () => {};
module.exports.bar = () => '';",
  "/node_modules/cjs-named-exports/package.json": "{
  \\"name\\": \\"cjs-named-exports\\"
}",
  "/node_modules/main-implicit/index.d.ts": "export = foo;
declare function foo(): void;",
  "/node_modules/main-implicit/index.js": "module.exports = function foo() {}",
  "/node_modules/main-implicit/package.json": "{
  \\"name\\": \\"main-implicit\\"
}",
  "/node_modules/typings-synonym/main-types.d.ts": "export function foo(): void;
export function bar(): string;",
  "/node_modules/typings-synonym/main.js": "module.exports.foo = () => {};
module.exports.bar = () => '';",
  "/node_modules/typings-synonym/package.json": "{
  \\"name\\": \\"typings-synonym\\",
  \\"main\\": \\"./main.js\\",
  \\"typings\\": \\"./main-types.d.ts\\"
}",
  "/package.json": "{
  \\"name\\": \\"untyped-cjs-and-esm\\",
  \\"main\\": \\"./main.js\\",
  \\"exports\\": {
    \\"./\\": {
      \\"require\\": \\"./main.js\\",
      \\"import\\": \\"./main.mjs\\"
    }
  }
}",
}
`;
